Improve CompletionKind DocC

rgoldberg opened the pull request:
Improve CompletionKind DocC.

Do not merge until after #738 has been merged.

Resolve #739

### Checklist
- [X] I've added at least one test that validates that my change is working, if appropriate
- [X] I've followed the code style of the rest of the project
- [X] I've read the [Contribution Guidelines](https://github.com/apple/swift-argument-parser/blob/main/CONTRIBUTING.md)
- [X] I've updated the documentation if necessary

natecook1000 commented on Sources/ArgumentParser/Parsable Properties/CompletionKind.swift:
````diff
@@ -9,56 +9,225 @@
+///   unsetopt aliases banghist
+///   ```
 public struct CompletionKind {
+  /// The type of completion to use for an argument or option value.
````
We can skip the doc comments on the enum and its cases, since they're all internal.

natecook1000 commented on Sources/ArgumentParser/Parsable Properties/CompletionKind.swift:
```diff
@@ -9,56 +9,225 @@
   }
 
-  /// Use the specified list of completion strings.
+  /// The completion candidates are the elements of `words`.
```
When it sounds natural, we try to just describe the argument in English words rather than using types or parameter names — something like:

```suggestion
  /// The completion candidates are the strings in the given array.
```

natecook1000 commented on Sources/ArgumentParser/Parsable Properties/CompletionKind.swift:
```diff
@@ -9,56 +9,225 @@
   }
 
-  /// Complete file names.
+  // swift-format-ignore: BeginDocumentationCommentWithOneLineSummary
+  /// The completion candidates include:
```
The one-line summary is what shows up as an abstract when an API is included as a list, so this would just be a sentence fragment. Let's use a full sentence ("...directory and file names, filtered by the given list of extensions"?) and then include the detail about exactly which file names in the body of the doc comment.

natecook1000 commented on Sources/ArgumentParser/Parsable Properties/CompletionKind.swift:
```diff
@@ -9,56 +9,225 @@
+  /// The completion candidates are directory names.
+  ///
+  /// The directory filter is included in a completion script when it is
+  /// generated.
```
I like these cues about how/where the different completion configurations are used 👍🏻

natecook1000 reviewed (commented):
These documentation additions are great, @rgoldberg – thanks so much! A few notes on the specifics below…

rgoldberg commented:
@natecook1000 I've addressed all your comments.

Unfortunately, I was lax on this PR like I was on the fish PR; I forgot that there are a few things that I must check first, then update the documentation (& possibly the code). I have therefore switched this PR to a draft, and I've included `TODO` notes in the DocC in such places.

I also included in this PR the `CompletionKind.custom` DocC improvements from my forthcoming PR that adds the 2 index parameters to the custom completion closures that aren't directly related to the 2 new index arguments. This significantly improves that DocC, but also significantly lengthens it with specificity & with exceptions to the standard behavior. Feel free to look over everything to provide feedback, or feel free to wait until I've fixed up all the TODOs (at which time I will mark the PR as ready for review). I'll finish the DocC here only after I've fixed up the fish PR.

rgoldberg commented:
@natecook1000 After looking through the possible zsh `setopt` settings, I want to change a few for the zsh scripts.

Should I create a new PR for the zsh `setopt` changes, then rebase this PR on `main` after that PR has been merged into `main`, or can I just make the zsh `setopt` settings in this PR?

The latter has less overhead, but munges a few glob settings into a documentation PR.

I want to follow whatever procedure you prefer (in the future, I'll default to whichever you prefer: splitting things up vs. munging small things together when convenient).

rgoldberg commented:
@natecook1000 If you have time, can you advise on the issues mentioned in the comment above? https://github.com/apple/swift-argument-parser/pull/740#issuecomment-2668651626

Would be good for me to test / finish this while you're reviewing my fish PR after that's ready.

natecook1000 commented on Sources/ArgumentParser/Parsable Properties/CompletionKind.swift:
```diff
@@ -9,56 +9,162 @@
+  /// order as in the command line. Note that shell words may contain spaces if
+  /// they are escaped or quoted.
+  ///
+  /// TODO: ensure that process substitutions work in bash 4. If not, check bash
```
There's enough here that it would be good to use subheads for the shell-specific notes, once the discussion that applies across the board is finished:

```
### Bash

In bash 3?-, a process substitution...

### Fish

In fish 3-, due to a bug...

etc
```

rgoldberg replied:
@natecook1000 I missed this comment before.  I'll include the subheadings now.

natecook1000 commented:
#### Re: Splitting PRs vs. munging very small issues together

I'm honestly happy to work with whatever makes sense to you – the separate PRs for different shell improvements have been nice to work with, and since you're taking over the effort to base completion generation on the tool info, there isn't much contention around this part of the code base. Why don't you go with what makes sense, and if something looks unreasonably large we can talk about how to split it up?

#### Postponing other shell settings

Agreed on all counts - happy for you to focus on the issues that are more pressing!

#### Bash 4

I don't know of any binary distributions, and it looks like I only have easy access to versions 3 and 5 — I'll ask around! Not a dealbreaker if we have to go without testing this version, since it seems unlikely to exist that frequently in the wild.

rgoldberg commented:
@natecook1000 Thanks for the feedback.

I'll probably munge the settings changes into the documentation PR, because they're very small, and I will probably not be able to work on SAP much right now, so I'd rather save time instead of coordinating multiple PRs, rebasing, etc.

The main interest I have in bash 4 is to ensure that the DocC is correct & concise, which is easier if I know when bash behavior changed (I'll probably check only bash 4 & 5, not minor versions or smaller increments).

Thanks for the help.

rgoldberg commented:
@natecook1000 Did you have any luck finding a binary of bash 4? Or any way in which I can easily compile it myself? I just need it to test some features to ensure that my revised documentation is correct.

FYI: I have an M4 Mac Studio now, so I have more build options as I am running the latest macOS & Xcode, and can run some other software that my old Mac mini 2014 couldn't. I still have my Mac mini 2014, though, so I can still run or build x86_64 binaries natively, if you only have binaries or working build instructions for x86_64.

Thanks for any help.

natecook1000 commented:
Still no luck on the bash 4 binary! Given how hard it is to find, I think it's fine to just do our best effort here and revise if we're able to get feedback down the line.

rgoldberg commented:
@natecook1000 Thanks for looking. I'll submit a new, non-draft, version of this PR in a few minutes after I've rebased it on current main.

rgoldberg commented:
@natecook1000 Please let me know if you want any changes to this.

natecook1000 commented on Sources/ArgumentParser/Parsable Properties/CompletionKind.swift:
```diff
@@ -9,56 +9,149 @@
   }
 
-  /// Complete file names.
+  /// The completion candidates include directory & file names, the latter
```
```suggestion
  /// The completion candidates include directory and file names, the latter
```

natecook1000 commented on Sources/ArgumentParser/Parsable Properties/CompletionKind.swift:
```diff
@@ -9,56 +9,149 @@
   }
 
-  /// Call the given shell command to generate completions.
+  /// The completion candidates are specified by the stdout output of the given
```
```suggestion
  /// The completion candidates are specified by the `stdout` output of the given
```

natecook1000 commented on Sources/ArgumentParser/Parsable Properties/CompletionKind.swift:
```diff
@@ -9,56 +9,149 @@
+  /// Given file extensions are parsed by the requesting shell as globs; Swift
+  /// Argument Parser does not perform any escaping or quoting.
+  ///
+  /// The directory/file filter & the given list of extensions are included in a
```
```suggestion
  /// The directory/file filter and the given list of extensions are included in a
```

natecook1000 commented on Sources/ArgumentParser/Parsable Properties/CompletionKind.swift:
```diff
@@ -9,56 +9,149 @@
   }
 
-  /// Generate completions using the given closure.
+  /// The completion candidates are the strings in the array returned by the
+  /// given closure when it is run when a user requests completions.
```
How about:
```suggestion
  /// The completion candidates are the strings in the array returned by the
  /// given closure when it is executed in response to a user's request for 
  /// completions.
```

natecook1000 commented on Sources/ArgumentParser/Parsable Properties/CompletionKind.swift:
```diff
@@ -9,56 +9,149 @@
+  /// order as in the command line. Note that shell words may contain spaces if
+  /// they are escaped or quoted.
+  ///
+  /// Shell words are passed to Swift verbatim, not unquoted. e.g., the
+  /// representation in Swift of the shell word `"abc\\""def"` would be exactly
+  /// the same, including the double quotes & the double backslash.
```
I think I know what you mean by "not unquoted", but the adjacent double negative is hard to parse. Can you expand this paragraph a bit? Let's also spell out "For example" instead of "e.g.".

natecook1000 reviewed (commented):
This is great! Some mostly minor notes below.

rgoldberg commented:
@natecook1000 I just pushed updated DocC. Please let me know if you want any more changes. Thanks.

rgoldberg commented:
@natecook1000 FYI: a fish dev is [working on the redirect issue](https://github.com/fish-shell/fish-shell/issues/11084#issuecomment-2851200447), but I don't know how soon it will be ready. I haven't looked at his branch yet. I assume that SAP should release 1.6.0 regardless of the fish progress; if I need to modify the SAP fish completions to work with the fish fix, that can be released in a future SAP version.

natecook1000 commented:
@swift-ci Please test
